core/src/main/java/com/google/errorprone/bugpatterns/AssertFalse.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.bugpatterns;
+
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker.AssertTreeMatcher;
+import com.google.errorprone.fixes.Fix;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+
+import com.sun.source.tree.AssertTree;
+import com.sun.source.tree.ExpressionTree;
+
+import static com.google.errorprone.BugPattern.Category.JDK;
+import static com.google.errorprone.BugPattern.MaturityLevel.EXPERIMENTAL;
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.matchers.Matchers.kindIs;
+import static com.sun.source.tree.Tree.Kind.BOOLEAN_LITERAL;
+import static com.sun.tools.javac.tree.JCTree.JCLiteral;
+
+/**
+ * @author sebastian.h.monte@gmail.com (Sebastian Monte)
+ */
+@BugPattern(name = "AssertFalse",
+    summary = "Assert false should not be used",
```

Comment:
The summary and explanation fields are presented to the user as the compile error message and documentation page content on errorprone.info, respectively.  We try to be as specific and clear as possible so that people understand why we're flagging the error and how to fix it.  Otherwise people tend to assume that the tool is wrong and suppress the error instead of fixing it.

For summary, I would say something like, "Assertions may be disabled at runtime and do not guarantee that execution will halt here; consider throwing an exception instead."  For explanation, something like, "Java assertions do not necessarily execute at runtime; they may be enabled and disabled depending on which options are passed to the JVM invocation. An `assert false` statement may be intended to ensure that the program never proceeds beyond that statement.  If the correct execution of the program depends on that being the case, consider throwing an exception instead, so that execution is halted regardless of runtime configuration."